### PR TITLE
SVE: Fix Shops Away from Shops, Junimo Woods Support, Lacey + Boarding House Support

### PR DESCRIPTION
**Pull request by Witchybun**
First half:
- Fix Gus' shop when he's on the island.
- Removes recipes from Marlon when on the phone.

The second half changes how four of the six junimo behave in the woods:
- Blue Junimo sells any fish caught in season.
- Grey Junimo sells any mineral donated to the museum.
- Red Junimo sells any artifact donated to the museum.
- Yellow Junimo sells any seeds that Pierre would likely sell to the player.

The caveat is that junimo don't usually deal in money, so these shops deal in items relevant to the color of the junimo.
- Exchange value is determined by finding the least common multiple (LCM) of the price of the two items to determine the perfect exchange rate between the two items.
- If the exchange rate's smallest value is larger than 5, the closest power of 5 is used to reduce this smallest value to be 5 or lower.
- Once done there is cleanup to ensure the ratio values are relatively prime (to reduce chance of a redundant trade).

The third half adds Hat Mouse Lacey and Boarding House support:
- Includes entrance randomizer support for Hat Mouse Lacey, name dictionary for NPC name.
- Includes entrance randomizer support for Boarding House, adds Gregory to late NPC.
- Includes location code for the chests you find on each abandoned mine map, up to five.

### CODE REVIEW

File: StardewArchipelago/Locations/CodeInjections/Modded/MagicModInjections.cs [hunk 1]
```diff
@@ -107,36 +114,51 @@ public static bool OnCast_AnalyzeGivesLocations_Prefix(object __instance, Farmer
...
+                {
+                    _shopReplacer.PlaceShopRecipeCheck(shopMenu.itemPriceAndStock, $"{recipe.ItemName} Recipe", recipe.ItemName, myActiveHints, recipe.Price);
                 }
+            }
+        }
 
-                _lastShopMenuUpdated = __instance;
-                if (!_archipelago.SlotData.Craftsanity.HasFlag(Craftsanity.All))
+
+        public static void Update_ReplaceMarlonShopChecks_Postfix(ShopMenu __instance, GameTime time)
```
Comment (agilbert1412) [inline StardewArchipelago/Locations/CodeInjections/Modded/MagicModInjections.cs:139]: Please add the signature of the function you're patching above the patch like in other places

File: StardewArchipelago/Locations/CodeInjections/Modded/SVE/SVEShopInjections.cs [hunk 2]
```diff
@@ -38,6 +38,7 @@ public class SVEShopInjections
             { new ShopIdentification("FishShop"), new[] { new PricedItem("Seaweed Salad", 1250) } },
             { new ShopIdentification("Sewer"), new[] { new PricedItem("Void Delight", 5000), new PricedItem("Void Salmon Sushi", 5000) } },
             { new ShopIdentification("ResortBar"), new[] { new PricedItem("Big Bark Burger", 5500), new PricedItem("Glazed Butterfish", 4000), new PricedItem("Mixed Berry Pie", 3500) } },
+            { new ShopIdentification("FarmHouse", "Marlon"), new[] { new PricedItem("Frog Legs", 2000), new PricedItem("Mushroom Berry Rice", 1500) } },
```
Comment (agilbert1412) [inline StardewArchipelago/Locations/CodeInjections/Modded/SVE/SVEShopInjections.cs:41]: I don't think we should sell these from the phone?

File: StardewArchipelago/Locations/CodeInjections/Modded/MagicModInjections.cs [hunk 3]
```diff
@@ -51,6 +52,12 @@ public static class MagicModInjections
         private static ShopReplacer _shopReplacer;
         private static ShopMenu _lastShopMenuUpdated = null;
 
+        private static readonly Dictionary<ShopIdentification, PricedItem[]> craftsanityRecipes = new()
+        {
+            { new ShopIdentification("AdventureGuild", "Marlon"), new[] { new PricedItem("Magic Elixir", 3000), new PricedItem("Travel Charm", 250) } },
+            { new ShopIdentification("FarmHouse", "Marlon"), new[] { new PricedItem("Magic Elixir", 3000), new PricedItem("Travel Charm", 250) } },
```
Comment (agilbert1412) [inline StardewArchipelago/Locations/CodeInjections/Modded/MagicModInjections.cs:58]: I don't think we should sell these from the phone?

**Review by agilbert1412 — CHANGES_REQUESTED** [review]

**Comment by agilbert1412** [comment]
You can't say "Self Explanatory" as a description when the PR contains a bunch of Marlon-related changes that are not in the title lmao

**Comment by Witchybun** [comment]
I renamed it due to the comment; it was named before I noticed Marlon with the same problem.  That's it.

**Comment by Witchybun** [comment]
I think I will just clear my workspace.  Have fun with this lmao

**Comment by Witchybun** [comment]
I'll clean this up and just send it later.

**Pull request closed without merging**